### ISSUE TITLE
Sequential search of ROM class in Anonymous classes unloading

### ISSUE THREAD

**Issue by dmitripivkine** [eclipse/openj9#7790]
Anonymous classes unloading procedure uses sequential walk to find ROM class segment for each RAM class to be unloaded. It can be reason for performance degradation if number of Anonymous classes is large. Also with adding Anonymous classes support to SC searched ROM class can not be in the Anonymous classloader list at all, so it is worst case scenario

**Comment by dmitripivkine** [comment]
We can use `findMemorySegment()` to search in `classMemorySegments` list. It would be a win for relatively large number of Anonymous classes in dedicated Anonymous Classloader. It would prevents n**2 search time obviously. However it might be slower for small number of Anonymous classes.
@tajila @DanHeidinga

**Comment by dmitripivkine** [comment]
An alternative suggestion is to keep pointer to ROM class segment from RAM class segment or J9Class itself. This will allow us to keep unloading procedure linear

**Comment by dmitripivkine** [comment]
@hangshao0 If ROM class for Anonymous class came from SC, would it be added to classMemorySegments list? Is there a simple way to recognize such ROM class?

**Comment by DanHeidinga** [comment]
> An alternative suggestion is to keep pointer to ROM class segment from RAM class segment or J9Class itself. This will allow us to keep unloading procedure linear

This would work for anonymous classes as they are 1-1 with their ram to rom segments.  We'd need to add the mapping in a way that doesn't add a cost to non-anon segments as they are not strictly 1-1

**Comment by hangshao0** [comment]
> If ROM class for Anonymous class came from SC, would it be added to classMemorySegments list? Is there a simple way to recognize such ROM class?

Yes, it will be in the classMemorySegments list. You can use j9shr_Query_IsAddressInCache() to tell if a romClass is from shared cache or not.
